prow/splice/main.go

```diff
@@ -0,0 +1,227 @@
+	remoteURL      = flag.String("remote-url", "https://github.com/kubernetes/kubernetes", "Remote Git URL")
+	githubRepoRE   = regexp.MustCompile(`/([^/]*/[^/]*)`)
+)
+
+// Call a binary and return its output and success status.
+func call(binary string, args ...string) (string, error) {
+	fmt.Print("+ ", binary, " ")
+	for _, arg := range args {
+		fmt.Print(arg, " ")
+	}
+	fmt.Print("\n")
+
+	cmd := exec.Command(binary, args...)
+	output, err := cmd.CombinedOutput()
+	return string(output), err
+}
+
+// getQueuedPRs reads the list of queued PRs from the Submit Queue.
+func getQueuedPRs(url string) ([]int, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+
+	queue := struct {
+		E2EQueue []struct {
+			Number int
+		}
+	}{}
+	err = json.Unmarshal(body, &queue)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := []int{}
+	for _, e := range queue.E2EQueue {
+		ret = append(ret, e.Number)
+	}
+	return ret, nil
+}
+
+// Splicer manages a git repo in specific directory.
+type splicer struct {
+	dir string // The repository location.
+}
+
+// makeSplicer returns a splicer in a new temporary directory,
+// with an initial .git dir created.
+func makeSplicer() (*splicer, error) {
+	dir, err := ioutil.TempDir("", "splice_")
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println(dir)
+	s := &splicer{dir}
+	err = s.gitCalls([][]string{
+		{"init"},
+		{"config", "--local", "user.name", "K8S Prow Splice"},
+		{"config", "--local", "user.email", "splice@localhost"},
+	})
+	if err != nil {
+		s.cleanup()
+		return nil, err
+	}
+	return s, nil
+}
+
+// cleanup recurisvely deletes the repository
+func (s *splicer) cleanup() {
+	os.RemoveAll(s.dir)
+}
+
+// gitCall is a helper to call `git -C $path $args`.
+func (s *splicer) gitCall(args ...string) error {
+	fullArgs := append([]string{"-C", s.dir}, args...)
+	output, err := call("git", fullArgs...)
+	fmt.Print(output)
+	return err
+}
+
+// gitCalls is a helper to chain repeated gitCall invocations,
+// returning the first failure, or nil if they all succeeded.
+func (s *splicer) gitCalls(argsList [][]string) error {
+	for _, args := range argsList {
+		err := s.gitCall(args...)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// findMergeable fetches given PRs from upstream, merges them locally,
+// and finally returns a list of PRs that can be merged without conflicts.
+func (s *splicer) findMergeable(remote string, prs []int) ([]int, error) {
+	args := []string{"fetch", remote, "master:master"}
+	for _, pr := range prs {
+		args = append(args, fmt.Sprintf("pull/%d/head:pr/%d", pr, pr))
+	}
+
+	err := s.gitCalls([][]string{
+		{"reset", "--hard"},
+		{"checkout", "--orphan", "blank"},
+		{"reset", "--hard"},
+		{"clean", "-fdx"},
+		args,
+		{"checkout", "-B", "batch", "master"},
+	})
+	if err != nil {
```

Comment:
I'm actually OK with the fetch failing occasionally here, since any errors in the main loop just cause it to sleep for a minute. It keeps the logic simpler to abort quickly.